### PR TITLE
use small lock 

### PR DESCRIPTION
**Pull request by hujun260**
## Summary

use small lock in:
arch/arm/src/imxrt/imxrt_wdog.c
arch/arm/src/kinetis/kinetis_edma.c
arch/arm/src/lc823450/lc823450_dvfs2.c
arch/arm/src/lc823450/lc823450_timer.c
arch/arm/src/lpc54xx/lpc54_lowputc.c

## Impact
none

## Testing
ci

### CODE REVIEW

**Comment by nuttxpr** [comment]
[**\[Experimental Bot, please feedback here\]**](https://github.com/search?q=repo%3Aapache%2Fnuttx+13552&type=issues)



Yes, this PR appears to meet the basic NuttX requirements, but it's missing crucial details.  While it provides a summary of *what* changed, it lacks the *why*.  The impact section is too brief, and the testing section is insufficient.

Here's a breakdown of what's missing and how to improve it:

* **Summary:**  Needs to explain the rationale behind using smaller locks.  What problem does this solve?  Is it a performance optimization?  A bug fix?  Are there deadlocks being addressed?  Be explicit.  Also, specify the issue number if one exists.

* **Impact:** While stating "none" is concise, it's likely inaccurate.  Consider these points:

    * **Performance:** Does using smaller locks improve performance?  If so, quantify it if possible. Even a qualitative assessment (e.g., "Slight performance improvement expected") is better than nothing.
    * **Memory usage:** Do the smaller locks reduce memory usage?
    * **Real-time behavior:**  Does this change affect real-time guarantees?

    If the answer truly is "no impact" in all these areas, explicitly state it for each point (e.g., "Impact on performance: NO").

* **Testing:**  "ci" is not sufficient.  While CI passing is important, the PR needs to show *specific* tests performed locally to validate the change.  This includes:

    * **Build Host(s):**  Specify the OS, CPU architecture, and compiler used.
    * **Target(s):** List the specific architectures and boards tested.  "sim" is not enough; specify which simulator.
    * **Logs:** Provide relevant logs *before* and *after* the change to demonstrate the improvement or fix.  If the change is a performance improvement, include timing measurements.  If it's a bug fix, show the erroneous behavior before and the corrected behavior after.

**Example of an improved Testing section:**

```
Testing

I confirm that changes are verified on local setup and works as intended:

* Build Host(s): Linux (Ubuntu 20.04), x86_64, GCC 9.4.0
* Target(s):
    * sim:qemu_armv7m (imxrt1064-evk config)
    * ARM:  STM32F429I-DISC1 (default config)


Testing logs before change (imxrt1064-evk):

```
[log output showing previous behavior, e.g., timing measurements or error messages]
```

Testing logs after change (imxrt1064-evk):

```
[log output showing improved behavior, e.g., faster timing or absence of errors]
```
```


By providing more detail and specific evidence, you'll make the review process smoother and increase the chances of your PR being accepted.

File: arch/arm/src/lpc54xx/lpc54_lowputc.c [hunk 1]
```diff
@@ -797,18 +799,18 @@ void arm_lowputc(char ch)
...
           /* Send the character */
 
           putreg32((uint32_t)ch, CONSOLE_BASE + LPC54_USART_FIFOWR_OFFSET);
-          spin_unlock_irqrestore(NULL, flags);
+          spin_unlock_irqrestore(&g_console_lock, flags);
           return;
         }
 
-      spin_unlock_irqrestore(NULL, flags);
+      spin_unlock_irqrestore(&g_console_lock, flags);
```
Comment (anchao) [inline arch/arm/src/lpc54xx/lpc54_lowputc.c:813]:
```
  irqstate_t flags = spin_lock_irqsave(&g_console_lock);

  while ((getreg32(CONSOLE_BASE + LPC54_USART_FIFOSTAT_OFFSET) &
        USART_FIFOSTAT_TXNOTFULL) == 0);

  putreg32((uint32_t)ch, CONSOLE_BASE + LPC54_USART_FIFOWR_OFFSET);

  spin_unlock_irqrestore(&g_console_lock, flags);

```

File: arch/arm/src/lc823450/lc823450_timer.c [hunk 2]
```diff
@@ -148,6 +148,8 @@ static void hrt_usleep_add(struct hrt_s *phrt);
  * Private Data
  ****************************************************************************/
 
+static spinlock_t g_rtc_lock = SP_UNLOCKED;
```
Comment (anchao) [inline arch/arm/src/lc823450/lc823450_timer.c:151]:
```suggestion
static spinlock_t g_hrt_timer_queue_lock = SP_UNLOCKED;
```
and move to line 141

File: arch/arm/src/kinetis/kinetis_edma.c [hunk 3]
```diff
@@ -197,15 +199,16 @@ static struct kinetis_edmatcd_s *kinetis_tcd_alloc(void)
    * waiting.
    */
 
-  flags = enter_critical_section();
   nxsem_wait_uninterruptible(&g_edma.dsem);
 
+  flags = spin_lock_irqsave(&g_edma_lock);
+
```
Comment (anchao) [inline arch/arm/src/kinetis/kinetis_edma.c:205]: from nxsem_wait_uninterruptible to sq_remfirst must be in atomic region, we can not replace spinlock arbitrarily

File: arch/arm/src/lc823450/lc823450_timer.c (hunk 2 above)
Comment (hujun260) [reply arch/arm/src/lc823450/lc823450_timer.c]: done

File: arch/arm/src/lpc54xx/lpc54_lowputc.c (hunk 1 above)
Comment (hujun260) [reply arch/arm/src/lpc54xx/lpc54_lowputc.c:801]: done

File: arch/arm/src/lpc54xx/lpc54_lowputc.c [hunk 4]
```diff
@@ -788,27 +790,15 @@ void arm_lowputc(char ch)
     {
```
Comment (anchao) [inline arch/arm/src/lpc54xx/lpc54_lowputc.c:790]: remove if ....

File: arch/arm/src/kinetis/kinetis_edma.c [hunk 5]
```diff
@@ -150,6 +150,8 @@ struct kinetis_edma_s
  * Private Data
  ****************************************************************************/
 
+static spinlock_t g_edma_lock = SP_UNLOCKED;
```
Comment (xiaoxiang781216) [inline arch/arm/src/kinetis/kinetis_edma.c:153]: move to kinetis_edma_s

File: arch/arm/src/imxrt/imxrt_wdog.c [hunk 6]
```diff
@@ -395,9 +398,11 @@ void imxrt_wdog_disable_all(void)
...
-  flags = enter_critical_section();
+  flags = spin_lock_irqsave(&g_wdgdev.lock);
   putreg32(RTWDOG_UPDATE_KEY, IMXRT_RTWDOG_CNT);
   putreg32(0xffff, IMXRT_RTWDOG_TOVAL);
   modifyreg32(IMXRT_RTWDOG_CS, RTWDOG_CS_EN, RTWDOG_CS_UPDATE);
-  leave_critical_section(flags);
+  spin_unlock_irqrestore(&g_wdgdev.lock, flags);
 }
+
+#endif /* CONFIG_WATCHDOG && CONFIG_IMXRT_WDOG */
```
Comment (xiaoxiang781216) [inline arch/arm/src/imxrt/imxrt_wdog.c:408]: why change

File: arch/arm/src/kinetis/kinetis_edma.c (hunk 5 above)
Comment (hujun260) [reply arch/arm/src/kinetis/kinetis_edma.c]: done

File: arch/arm/src/kinetis/kinetis_edma.c [hunk 7]
```diff
@@ -1162,14 +1178,11 @@ int kinetis_dmach_start(DMACH_HANDLE handle, edma_callback_t callback,
 void kinetis_dmach_stop(DMACH_HANDLE handle)
 {
   struct kinetis_dmach_s *dmach = (struct kinetis_dmach_s *)handle;
-  irqstate_t flags;
```
Comment (xiaoxiang781216) [inline arch/arm/src/kinetis/kinetis_edma.c:1165]: why remove the lock

File: arch/arm/src/kinetis/kinetis_edma.c [hunk 8]
```diff
@@ -228,10 +230,24 @@ static void kinetis_tcd_free(struct kinetis_edmatcd_s *tcd)
...
+{
+  irqstate_t flags;
+
+  /* Add the the TCD to the end of the free list and post the 'dsem',
+   * possibly waking up another thread that might be waiting for
+   * a TCD.
+   */
+
+  flags = spin_lock_irqsave(&g_edma.lock);
+  sched_lock();
```
[the remark this answers is not in the document]
Comment (hujun260) [reply arch/arm/src/kinetis/kinetis_edma.c:247]: I believe that disabling irq before disabling scheduling can improve system efficiency in some cases.
If scheduling is disabled first and then irq are disabled, an interrupt may occur within this extremely short time window,
but scheduling cannot be performed at this time, resulting in a wasted context switch.
The implementation of spin_lock_irqsave in Linux also follows the order of disabling irq first and then disabling scheduling.

File: arch/arm/src/kinetis/kinetis_edma.c (hunk 7 above)
Comment (hujun260) [reply arch/arm/src/kinetis/kinetis_edma.c:1165]: I have implemented it in kinetis_dmaterminate.

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**